docker/driver.go

```diff
@@ -0,0 +1,99 @@
+package docker
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/bblfsh/performance"
+
+	"bitbucket.org/creachadair/shell"
+	"github.com/creachadair/staticfile"
+	"gopkg.in/src-d/go-errors.v1"
+	"gopkg.in/src-d/go-log.v1"
+)
+
+//go:generate go run github.com/creachadair/staticfile/compiledata -pkg docker -out static.go conf/drivers.json
+
+// InstallDriver downloads driver's repository, checkouts to a given commit,
+// builds docker image and installs driver's image to bblfshd container.
+// Requires bblfshd container running.
+func InstallDriver(language, commit string) (func(), error) {
```

Comment:
I suggest simplifying the cleanup interface here in the following way:

1. Ensure that when an error is returned, the cleanup has already been done, and
2. Ensure that when no error is returned, the caller _always_ has a valid function to call.

That way the caller doesn't need to do a fiddly check for `nil`, and the semantics of cleanup are all in one place: If they get a non-nil error, they have to call the cleanup (or explicitly choose not to):

```go
cleanup, err := docker.InstallDriver("foo", "bar")
if err != nil {
   log.HaltAndCatchFire(err)
}
defer cleanup()
doOtherStuffWith("foo", "bar")
…
```

I think the only changes that would need to happen for this are to call the cleanup function explicitly in the fail branches of the error checks at 42 & 45. Maybe not even then, if `downloadAndBuildDriver` cleans up after its own errors.

Reply:
Done

Reply:
I think there are still some missing pieces: Some paths return a cleanup function, some return nil (

I think it's OK for us to either _always_ return a valid cleanup function—which I think was the intent here—or to only ever return one when `err == nil`. We should choose one or the other, though, and be consistent.

Reply:
Decide not to return cleanup at all, please check it.